java-checks/src/main/java/org/sonar/java/checks/OverwrittenKeyCheck.java

```diff
@@ -0,0 +1,177 @@
+
+    private CollectionAndKey(Symbol collection, Tree keyTree, Object key) {
+      this.collection = collection;
+      this.keyTree = keyTree;
+      this.key = key;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      CollectionAndKey that = (CollectionAndKey) o;
+      return Objects.equals(collection, that.collection) &&
+        Objects.equals(key, that.key);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(collection, key);
+    }
+  }
+
+  @CheckForNull
+  private static Symbol symbolFromIdentifier(ExpressionTree collectionExpression) {
+    if (collectionExpression.is(Tree.Kind.IDENTIFIER)) {
+      return ((IdentifierTree) collectionExpression).symbol();
+    }
+    return null;
+  }
+
+  @CheckForNull
+  private static CollectionAndKey isArrayAssignment(StatementTree statementTree) {
+    if (statementTree.is(Tree.Kind.EXPRESSION_STATEMENT)) {
+      ExpressionTree expression = ((ExpressionStatementTree) statementTree).expression();
+      if (expression.is(Tree.Kind.ASSIGNMENT)) {
+        ExpressionTree variable = ((AssignmentExpressionTree) expression).variable();
+        if (variable.is(Tree.Kind.ARRAY_ACCESS_EXPRESSION)) {
+          ArrayAccessExpressionTree aaet = (ArrayAccessExpressionTree) variable;
+          Symbol collection = symbolFromIdentifier(aaet.expression());
+          ExpressionTree keyTree = aaet.dimension().expression();
+          Object key = extractKey(keyTree);
+          if (collection != null && key != null) {
+            return new CollectionAndKey(collection, keyTree, key);
+          }
+        }
+      }
+    }
+    return null;
+  }
+
+  @CheckForNull
+  private static CollectionAndKey isMapPut(StatementTree statementTree) {
+    if (statementTree.is(Tree.Kind.EXPRESSION_STATEMENT)) {
+      ExpressionTree expression = ((ExpressionStatementTree) statementTree).expression();
+      if (expression.is(Tree.Kind.METHOD_INVOCATION) && MAP_PUT.matches((MethodInvocationTree) expression)) {
+        MethodInvocationTree mapPut = (MethodInvocationTree) expression;
+        Symbol collection = mapPut.methodSelect().is(Tree.Kind.MEMBER_SELECT) ? symbolFromIdentifier(((MemberSelectExpressionTree) mapPut.methodSelect()).expression()) : null;
+        ExpressionTree keyTree = mapPut.arguments().get(0);
+        Object key = extractKey(keyTree);
+        if (collection != null && key != null) {
+          return new CollectionAndKey(collection, keyTree, key);
+        }
+      }
+    }
+    return null;
+  }
+
+  private void handleKey(Map<CollectionAndKey, Tree> keys, CollectionAndKey collectionAndKey) {
+    if (keys.containsKey(collectionAndKey)) {
+      Tree previousTree = keys.get(collectionAndKey);
+      String indexOrKey = collectionAndKey.keyTree.parent().is(Tree.Kind.ARRAY_DIMENSION) ? "index" : "key";
```

Comment:
I would encode the type of the collection into the Object, and put a method there, or precompute it when creating the object.